### ISSUE TITLE
New MSAL v0.8.0 breaks AZ Identity v1.2.0 in sdk for go

### ISSUE THREAD

**Issue by gcohler** [Azure/azure-sdk-for-go#19871]
### Bug Report

This is similar to #19823 but appears to be caused by a new release of MSAL.  Here is what I'm seeing, and my analysis / workaround.

Last night MSAL for go v0.8.0 was released.  Since that time, when I build my application, I get the following errors (see below).

If I force use of MSAL v0.7.0 by adding `go mod edit -require github.com/AzureAD/microsoft-authentication-library-for-go@v0.7.0` to my makefile, the problem goes away.  And my project builds as it always has.

So the work-around works.  But it appears that there is an incompatibility between `github.com/AzureAD/microsoft-authentication-library-for-go@v0.8.0` and `github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0`

I'm using go v1.19.5 -- but I also tried with v1.18 with the same results.

Errors when built with MSAL v0.8.0:

```
# github.com/Azure/azure-sdk-for-go/sdk/azidentity
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/client_assertion_credential.go:52:44: cannot use c (variable of type confidential.Client) as type confidentialClient in struct literal:
        confidential.Client does not implement confidentialClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...confidential.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...confidential.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/client_certificate_credential.go:59:46: cannot use c (variable of type confidential.Client) as type confidentialClient in struct literal:
        confidential.Client does not implement confidentialClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...confidential.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...confidential.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/client_secret_credential.go:43:41: cannot use c (variable of type confidential.Client) as type confidentialClient in struct literal:
        confidential.Client does not implement confidentialClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...confidential.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...confidential.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/device_code_credential.go:85:66: cannot use c (variable of type public.Client) as type publicClient in struct literal:
        public.Client does not implement publicClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...public.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...public.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/device_code_credential.go:94:59: cannot use public.WithSilentAccount(c.account) (value of type interface{public.AcquireSilentOption; options.CallOption}) as type public.AcquireTokenSilentOption in argument to c.client.AcquireTokenSilent
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/interactive_browser_credential.go:62:60: cannot use c (variable of type public.Client) as type publicClient in struct literal:
        public.Client does not implement publicClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...public.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...public.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/interactive_browser_credential.go:70:59: cannot use public.WithSilentAccount(c.account) (value of type interface{public.AcquireSilentOption; options.CallOption}) as type public.AcquireTokenSilentOption in argument to c.client.AcquireTokenSilent
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/interactive_browser_credential.go:78:17: cannot use public.WithRedirectURI(c.options.RedirectURL) (value of type interface{public.AcquireInteractiveOption; options.CallOption}) as type public.InteractiveAuthOption in argument to append
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/managed_identity_credential.go:101:44: cannot use c (variable of type confidential.Client) as type confidentialClient in struct literal:
        confidential.Client does not implement confidentialClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...confidential.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...confidential.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/username_password_credential.go:46:85: cannot use c (variable of type public.Client) as type publicClient in struct literal:
        public.Client does not implement publicClient (wrong type for AcquireTokenByAuthCode method)
                have AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, opts ...public.AcquireByAuthCodeOption) (base.AuthResult, error)
                want AcquireTokenByAuthCode(ctx context.Context, code string, redirectURI string, scopes []string, options ...public.AcquireTokenByAuthCodeOption) (base.AuthResult, error)
/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azidentity@v1.2.0/username_password_credential.go:46:85: too many errors
```

**Comment by chlowell** [comment]
This is due to the test fakes I mentioned in #19823. The next azidentity release will upgrade to the latest MSAL and update these fakes. In the meantime, continuing to use the prior `github.com/AzureAD/microsoft-authentication-library-for-go` version, v0.7.0, is the easiest workaround. If your application needs v0.8.0, [this commit](https://github.com/Azure/azure-sdk-for-go/pull/19529/commits/0e303ee6a3707dbcc8084f248852173a36eca8bd) shows how to make the azidentity test fakes compatible.

**Comment by chlowell** [comment]
Thanks for opening this issue! I'm closing it because we published azidentity v1.2.1 today, which is simply v1.2.0 with an upgrade to MSAL v0.8.1. For anyone using an azidentity v1.3.0 beta, the next beta release (coming soon) will include the same upgrade; you can get it sooner from our `main` branch: `go get -u github.com/Azure/azure-sdk-for-go/sdk/azidentity@main`.
